src/userobjects/MoabSkinner.C

```diff
@@ -0,0 +1,1008 @@
+
+    if (_verbose)
+      _console << "done" << std::endl;
+  }
+
+  if (_output_full)
+  {
+    std::string filename = "moab_mesh_" + extension;
+
+    if (_verbose)
+      _console << "Writing MOAB mesh to " << filename << std::endl;
+
+    _moab->write_mesh(filename.c_str());
+
+    if (_verbose)
+      _console << "done" << std::endl;
+  }
+
+  _n_write++;
+}
+
+void
+MoabSkinner::groupLocalElems(std::set<dof_id_type> elems, std::vector<moab::Range> & localElems)
+{
+  while (!elems.empty())
+  {
+
+    // Create a new local range of moab handles
+    moab::Range local;
+
+    // Retrieve and remove the fisrt elem
+    auto it = elems.begin();
+    dof_id_type next = *it;
+    elems.erase(it);
+
+    std::set<dof_id_type> neighbors;
+    neighbors.insert(next);
+
+    while (!neighbors.empty())
+    {
+
+      std::set<dof_id_type> new_neighbors;
+
+      // Loop over all the new neighbors
+      for (auto & next : neighbors)
+      {
+
+        // Get the MOAB handles, and add to local set
+        // (May be more than one if this libMesh elem has sub-tetrahedra)
+        if (_id_to_elem_handles.find(next) == _id_to_elem_handles.end())
+          mooseError("No entity handles found for libmesh id.");
+
+        std::vector<moab::EntityHandle> ents = _id_to_elem_handles[next];
+        for (const auto ent : ents)
+          local.insert(ent);
+
+        // Get the libMesh element
+        Elem & elem = mesh().elem_ref(next);
+
+        // How many nearest neighbors (general element)?
+        unsigned int NN = elem.n_neighbors();
+
+        // Loop over neighbors
+        for (unsigned int i = 0; i < NN; i++)
+        {
+          const Elem * nnptr = elem.neighbor_ptr(i);
+          // If on boundary, some may be null ptrs
+          if (nnptr == nullptr)
+            continue;
+
+          dof_id_type idnn = nnptr->id();
+
+          // Select only those that are in the current bin
+          if (elems.find(idnn) != elems.end())
+          {
+            new_neighbors.insert(idnn);
+            // Remove from those still available
+            elems.erase(idnn);
+          }
+        }
+      }
+
+      // Found all the new neighbors, done with current set.
+      neighbors = new_neighbors;
+    }
+
+    // Save this moab range of local neighbors
+    localElems.push_back(local);
+  }
+}
+
+void
+MoabSkinner::reset()
+{
+  if (!_fixed_mesh)
+  {
+    _moab.reset(new moab::Core());
+    skinner.reset(new moab::Skinner(_moab.get()));
+    gtt.reset(new moab::GeomTopoTool(_moab.get()));
+  }
+
+  // Clear entity set maps
+  surfsToVols.clear();
+}
+
+unsigned int
+MoabSkinner::getBin(const unsigned int & iVarBin,
+                       const unsigned int & iDenBin,
+                       const unsigned int & iMat) const
+{
+  return _n_temperature_bins * (_n_density_bins * iMat + iDenBin) + iVarBin;
+}
+
+unsigned int
+MoabSkinner::getMatBin(const unsigned int & iDenBin) const
```

Comment:
It's here for future capability. This is basically the function that is called to encode a region's material name, after the skinning operation. The function looks simpler than it is in Aurora, because in Aurora a new material is created even when only skinning by temperature (which isn't strictly needed, because you can change temperatures of cells filled with the same material without needing to duplicate the material).